core/src/main/java/org/elasticsearch/common/xcontent/json/JsonXContentGenerator.java

```diff
@@ -34,13 +41,50 @@
  */
 public class JsonXContentGenerator implements XContentGenerator {
 
-    protected final BaseJsonGenerator generator;
+    /** Generator used to write content **/
+    protected final JsonGenerator generator;
+
+    /**
+     * Reference to base generator because
+     * writing raw values needs a specific method call.
+     */
+    private final GeneratorBase base;
+
+    /**
+     * Reference to filtering generator because
+     * writing an empty object '{}' when everything is filtered
+     * out needs a specific treatment
+     */
+    private final FilteringGeneratorDelegate filter;
+
     private boolean writeLineFeedAtEnd;
     private static final SerializedString LF = new SerializedString("\n");
     private  static final DefaultPrettyPrinter.Indenter INDENTER = new DefaultIndenter("  ", LF.getValue());
 
-    public JsonXContentGenerator(BaseJsonGenerator generator) {
+    public JsonXContentGenerator(JsonGenerator generator) {
         this.generator = generator;
+        this.base = initGeneratorBase(generator);
+        this.filter = initFilteringGenerator(generator);
+    }
+
+    private GeneratorBase initGeneratorBase(JsonGenerator generator) {
+        if (generator instanceof GeneratorBase) {
+            return (GeneratorBase) generator;
+        }
+        if (generator instanceof JsonGeneratorDelegate) {
+            JsonGeneratorDelegate filter = (JsonGeneratorDelegate) generator;
+            if (filter.getDelegate() instanceof GeneratorBase) {
+                return (GeneratorBase) filter.getDelegate();
+            }
+        }
+        return null;
+    }
+
+    private FilteringGeneratorDelegate initFilteringGenerator(JsonGenerator generator) {
+        if (generator instanceof FilteringGeneratorDelegate) {
+            return (FilteringGeneratorDelegate) generator;
+        }
+        return null;
```

Comment:
I'm not too happy with those two methods. Could the constructor take those sub generators explicitly instead? Or maybe the constructor could take `JsonGenerator jsonGenerator, String... filters` and create them in the constructor?